gemrb/core/Scriptable/Actor.cpp

```diff
@@ -3463,38 +3463,46 @@ void Actor::UpdateFatigue()
 	if (!InParty || !game->GameTime) {
 		return;
 	}
-	// do icons here, so they persist for more than a tick
-	int LuckMod = core->ResolveStatBonus(this, "fatigue") ; // fatigmod.2da
-	if (LuckMod) {
-		AddPortraitIcon(PI_FATIGUE);
-	} else {
-		DisablePortraitIcon(PI_FATIGUE);
+
+	bool updated = false;
+	if (!TicksLastRested) {
+		// just loaded the game; approximate last rest
+		TicksLastRested = game->GameTime - (2*core->Time.hour_size) * (2*GetBase(IE_FATIGUE)+1);
+		updated = true;
+	} else if (LastFatigueCheck) {
+		ieDword FatigueDiff = (game->GameTime - TicksLastRested) / (4*core->Time.hour_size)
+		                    - (LastFatigueCheck - TicksLastRested) / (4*core->Time.hour_size);
+		if (FatigueDiff) {
+			NewBase(IE_FATIGUE, FatigueDiff, MOD_ADDITIVE);
+			updated = true;
+		}
+	}
+	LastFatigueCheck = game->GameTime;
+
+	if (!core->HasFeature(GF_AREA_OVERRIDE)) {
+		// pst has TNO regeneration stored there
+		// shouldn't we check for our own flag, though?
+		// FIXME: the Con bonus is applied dynamically, but this doesn't appear to conform to
+		// the original engine behavior?  we should probably apply the bonus on rest
+		int FatigueBonus = core->GetConstitutionBonus(STAT_CON_FATIGUE, Modified[IE_CON]);
+		if (Modified[IE_FATIGUE] >= FatigueBonus) {
+			Modified[IE_FATIGUE] -= FatigueBonus;
+		} else {
+			Modified[IE_FATIGUE] = 0;
+		}
 	}
 
-	ieDword FatigueLevel = (game->GameTime - TicksLastRested) / (4*core->Time.hour_size);
-	int FatigueBonus = core->GetConstitutionBonus(STAT_CON_FATIGUE, Modified[IE_CON]);
-	// pst has TNO regeneration stored there
-	if (core->HasFeature(GF_AREA_OVERRIDE)) FatigueBonus = 0;
-	FatigueLevel = (signed)FatigueLevel - FatigueBonus >= 0 ? FatigueLevel - FatigueBonus : 0;
-	FatigueLevel = ClampStat(IE_FATIGUE, FatigueLevel);
-
-	// don't run on init or we automatically make the character supertired
-	if (FatigueLevel != BaseStats[IE_FATIGUE] && TicksLastRested) {
-		int OldLuckMod = LuckMod;
-		NewBase(IE_FATIGUE, FatigueLevel, MOD_ABSOLUTE);
```

Comment:
Re: modified fatigue being recalculated, I can't find IE_FATIGUE being changed in Actor.cpp anywhere else but here before my fix.  And recall that `UpdateFatigue()` happens in the middle of `RefreshEffects()`, shortly after all the `Modified[]` is reset to `Base[]`.  And, importantly, `NewBase()` actually changes `Modified[]`, too.

Reply:
RefreshPCStats is ran after all the effects have been applied and Modified restored (save for any pcf changes — irrelevant for fatigue). NewBase makes sure to conserve the diff between BaseStats and Modified. Anyway, it's better to just test it out properly, which I already have on the todo.

Reply:
So, some numbers. The spell just adds 4 fatigue points, while 7 is the normal limit for feeling tired. Zelazko confirmed in the original that using it right after resting doesn't make you sleepy (Xzar with 10 Con and Viconia with 8 Con didn't get tired), so at least this is consistent and just the spell description is misleading.

Looking at the effect, it's of course delayed, but then doesn't stick around. Are you sure your version fixed haste for you? It should affect the base stat to have the desired effect, since it's not setting a good Modified for long.

Reply:
I'm back!  Yes, on further checks, Haste does not appear to affect fatigue at all, even after my fix.

But I think this might be because of another bug.  Look here:
https://github.com/gemrb/gemrb/blob/308010894dcf6e9fb811d36bfba6a4048564a79b/gemrb/core/EffectQueue.cpp#L209-L214
This array determines how delayed effect timing types transform after the delay expires. Is `FX_DURATION_DELAY_PERMANENT` really supposed to decay into `FX_DURATION_AFTER_EXPIRES`?  If I change it to `FX_DURATION_INSTANT_PERMANENT`, Haste fatigue starts to work (with my fix).

Reply:
I came to a similar conclusion — that either this or the opcode needs to change. But before changing something so central as fx_triggered, I'd like to see all other users, so probably need to write a short iesh script. Looks like NI can only look at externall eff files when searching, while dltcep can't do even that.

Reply:
Wrote the script and there are many effects with this timing mode in bg1. Most are just to play a sound on spell expiry, but here are the examples of each:

> SPIN103 | 174 0xae play sound (bhaal power Draw upon holy might)
> SPIN117 | 12  0xc damage
> SPIN117 | 139 0x8b display string
> SPIN996 | 151 0x97 ReplaceCreature
> SPPR205 | 146 0x92 Spell:Cast
> SPPR207 | 122 0x7a Item:CreateInventory
> SPWI402 | 124 0x7c DimensionDoor
> SPWI416 | 172 0xac Spell:Remove
> SPWI982 | 68  0x44 UnsummonCreature FX_APPLIED in second case

All but one case of the last opcode are one-off effects and they'd just pollute the effect queue if we kept them around. But your suggestion should still work, just didn't test if there are 2 correct applications. BG2 has more examples of which only ApplyEffect is problematic, since it can return any of the application modes by design. But it does seem like the fatigue opcode itself is fine and doesn't only work on the base stat.

Reply:
I think the effects won't be kept if they return `FX_NOT_APPLIED`, regardless of timing mode.  So, do I add the `fx_triggered` fix?  Or does it need more testing WRT `ApplyEffect`?

Reply:
I want to check the other games first, but so far it does seem like the change is fine. ApplyEffect should be fine too. In bg2 only these four spells use it with this timing mode: SPWI324, SPWI720, SPCL742, SPIN891. All are ok, 3/4 just using it to play an expiry sound.

Reply:
The rest looks fine too, not completely certain only about one very complicated pst opcode, but the green light is lit.

For posterity's sake, all the opcodes used with this timing mode:
> bg1 hits:
> SPIN117 | 12  0xc damage
> SPWI982 | 68  0x44 UnsummonCreature
> SPPR207 | 122 0x7a Item:CreateInventory
> SPWI402 | 124 0x7c DimensionDoor
> SPIN117 | 139 0x8b display string
> SPPR205 | 146 0x92 Spell:Cast
> SPIN996 | 151 0x97 ReplaceCreature
> SPWI416 | 172 0xac Spell:Remove
> SPIN103 | 174 0xae play sound
> 
> bg2
> SPIN952 | 0xd / 13 death
> SPWI305 | 0x5d / 93 fatigue PERMANENT
> SPIN916 | 0x8d / 141 VisualSpellHit
> SPIN846 | 0xa8 / 168  RemoveCreature
> SPIN952 | 0xaa / 170 DamageAnimation
> SPWI720 | 0xb1 / 177 ApplyEffect variable!
> SPWI910 | 0xd3 / 211  State:Imprisonment  
> SPWI813 | 0xd5 / 213 Maze 
> SPIN793 | 0xd7 / 215 PlayVisualEffect  variable!
> SPWI705 | 0xe6 / 230 DispelSecondaryTypeOne
> SPWI705 | 0xf0 / 240 Icon:Remove
> SPPR720 | 0x108 / 264 DropWeapon
> SPWI304 | 0x109 / 265 ModifyGlobalVariable
> 
> ApplyEffect bg2: SPWI324, SPWI720, SPCL742, SPIN891
> 
> pst extras:
> 0x1a / 26 | SPWI409 
> 0x6e / 110 | SPWI505 
> 0x8a / 138 | SPWI103 
> 0xbb / 187 | SPWI701 
> 0xbc / 188 | SPWI212 
> 0xbd / 189 | SPWI401
> 0xbf / 191 | SPWI905
> 0xc0 / 192 | SPIN104 
> 0xc1 / 193 | SPWI705 
> 0xc2 / 194 | SPWI409 
> 0xc4 / 196 | SPWI201 
> 0xc9 / 201 | SPWI704 ?
> 0xcf / 207 | SPWI308 
> 
> iwd2 extras:
> 0x42 / 66 | SPIN181
> 0x5d / 93 | SPIN136 fatigue +7
> 0xe8 / 232 | SPWI606
> 0xe9 / 233 | SPWI707
> 0x103 / 259 | SPIN985
> 0x104 / 260 | SPIN985
> 0x109 / 265 | SPPR409
> 0x113 / 275 | SPWI710
> 0x192 / 402 | SPWI713
> 0x19a / 410 | SPIN282
> 0x1a4 / 420 | SPWI606
> 
> iwd1 extras:
> 0x116 / 278 | SPWI511
> 0x4e / 78 | SPPR322